### ISSUE TITLE
How to define an option to accept only one value, but allow to use it multiple times?

### ISSUE THREAD

**Issue by RobertZenz** [remkop/picocli#190]
I can't figure out how to define an option which does accept only one value but can be specified multiple times. To be exact:

    application -o value1 -o value2 parameter1 parameter2

When defined without arity, `parameter1` and `parameter2` are added to `o`. I tried this:

	@Option(names = { "-o", "--open" }, arity = "1")
	private List<String> files = null;

But that disallows multiple usage of `-o`.

How can I define this?

**Comment by remkop** [comment]
If only one value is allowed, why not make it a String instead of a List?
```
@Option(names = { "-o", "--open" })
private String files = null;
```

Now, if the option is specified multiple times, picocli will throw a `OverwrittenOptionException` by default, but you can make it lenient by calling `CommandLine.setOverwrittenOptionsAllowed(true)` before parsing the input. 

This way the last option value specified on the command line is preserved. 

See also http://picocli.info/#_too_many_values

**Comment by RobertZenz** [comment]
Well, I want multiple values to be allowed, but only one to be specified to the option.

The idea is the following, I want to allow to have multiple options in front of an undefined number of parameters which will be concatenated into one.

    application -o first -o second para1 para2 para3

Which should yield the following:

    o = { "first", "second" }
    parameters = { "para1", "para2", "para3" }

Is that possible?

**Comment by remkop** [comment]
Oh sorry! I misread your original question. 

If you have an option that takes multiple values, picocli will keep assigning values to it until picocli encounters another option or the `--` end-of-options delimiter. 

So, input like this will be interpreted correctly:

```
application -o option1 -o option2 -- parameter1 parameter2
```

**Comment by RobertZenz** [comment]
Mh, I see what you mean.

Would there be a way to introduce an option to only have one value assigned to the option, something similar to the arity value? Maybe a "global" option on the parse or something that tells it to stop after one value has been encountered?

**Comment by remkop** [comment]
Once an option accepts multiple values, this kind of ambiguity is unfortunately unavoidable...
We are not the first to come across this issue, which is why the POSIX standard introduced the `--` end of options delimiter.

The thing is, the end user may specify the options flag only once, followed by multiple values:
```
application -o option1 option2 option3 param1 param2
```
If the `-o` field is a `List`, picocli should add all 5 specified values to the list.
The user needs to specify the `--` separator to help the parser disambiguate the input.

**Comment by RobertZenz** [comment]
Sorry, maybe I was not clear. I asked if there is an easy way to add an option to change that behavior. So, to switch from "add all values" to "add only one value".

**Comment by remkop** [comment]
So the desired behaviour is to have an option that can have multiple values, but the user needs to _explicitly specify the flag_ in front of each value. Each time the flag is specified, only one value is consumed. This would make user input unambiguous. I can see how this would be valuable.

Let's go through a few cases to make sure we have the same understanding.

Example 1: (flag specified multiple times, flag is explicitly specified in front of each value)
```
application -o option1 -o option2 param1 param2
```
Would result in 
```
o = { "option1", "option2" }
parameters = { "param1", "param2" }
```

Example 2: 
```
application -o option1 option2 -- param1 param2
```
Would result in 
```
o = { "option1" }
parameters = { "option2", "--", "param1", "param2" }
```

Example 3: 
```
application -o option1 -x XXX -o option2 param1 param2
```
Would result in 
```
o = { "option1", "option2" }
x = { "XXX" }
parameters = { "param1", "param2" }
```

Example 4: 
```
application -o option1 -x XXX YYY -o option2 param1 param2
```
The YYY value has no explicit flag in front of it, so it cannot be assigned to either the -x or the -o option. The remainder is parsed as positional parameters.  This would result in:
```
o = { "option1" }
x = { "XXX" }
parameters = { "YYY", "-o", "option2", "param1", "param2" }
```
(This last one is perhaps not desirable. There is a separate ticket (#130) for supporting a mixture of options and positional parameters on the command line. I wanted to mention the possibility, but let's keep this functionality out of the scope of this ticket.)


Then the question becomes how to control this behaviour. Perhaps an option attribute, like the below?
```
@Option(names="-o", explicitFlags=true)
```
`explicitFlags` is perhaps not a good name. Any ideas for how to name this attribute?

**Comment by RobertZenz** [comment]
Yes, that is exactly the behavior I'd expect (and want, in this case).

Regarding the flag, maybe providing the number of allowed values (similar to the arity value) is an option? That would provide the maximum customizability possible.

    @Option(names="-o", valuesCount="1")
    @Option(names="-o", valuesCount="1..*")
    @Option(names="-o", valuesCount="2..5")

**Comment by remkop** [comment]
My gut feeling is that that would be too confusing.

That would introduce two "number of allowed values": one for the max number of values that is stored in the List (arity), and another one for the number of values that may follow an occurrence of the option name on the command line.

If you need more flexibility you may want to use the `split` regex, so users can specify
```
application -o option1,option2,option3 param1 param2
```
...and if the option is annotated with `@Option(name="-o", split=",")`, picocli will consume a single command line argument (`option1,option2,option3`) and still put 3 separate values in the List.

As for naming, I would prefer to focus on the fact that users need to specify the option name repeatedly for each value to be added to the List. I think that is easier to explain and easier to understand.

**Comment by RobertZenz** [comment]
You're right, that level of detail is not needed (or already there in picocli). So my suggestions for your proposed flag would be something like `multipleValues` with defaulting to `true`. Or maybe `singleValue` which defaults to `false`. Maybe `optionRequired` or `explicitOption`?

**Comment by remkop** [comment]
Maybe `@Option(names="-o", greedy=false)`, where the default is `true`? The term "greedy" is fairly well-known because of its use in regular expressions, which behave similarly.

**Comment by RobertZenz** [comment]
That is a great idea! It's perfect.

**Comment by remkop** [comment]
Let me know if you feel like providing a pull request for this.

**Comment by RobertZenz** [comment]
I'll give it a try.

**Comment by RobertZenz** [comment]
Nope, sorry, I failed.

I've added an additional field to `Option` and then tried to add the logic to `consumeArguments(...)`, like this:

    private List<Object> consumeArguments(Field field,
                                          Class<?> annotation,
                                          Range arity,
                                          Stack<String> args,
                                          Class<?> type,
                                          int originalSize,
                                          String argDescription) throws Exception {
        List<Object> result = new ArrayList<Object>();
        int index = 0;

        // first do the arity.min mandatory parameters
        for (int i = 0; result.size() + originalSize < arity.min; i++) {
            index = consumeOneArgument(field, arity, args, type, result, index, originalSize, argDescription);
        }

        if (field.isAnnotationPresent(Option.class)) {
            if (!field.getAnnotation(Option.class).greedy()) {
                consumeOneArgument(field, arity, args, type, result, index, originalSize, argDescription);
                return result;
            }
        }

        // now process the varargs if any
        while (result.size() < arity.max && !args.isEmpty()) {
            if (annotation != Parameters.class) {
                if (commands.containsKey(args.peek()) || isOption(args.peek())) {
                    return result;
                }
            }
            index = consumeOneArgument(field, arity, args, type, result, index, originalSize, argDescription);
        }
        return result;
    }

Which seems to work. But then I found that `processArguments` is finding one unmatched parameter and treats everything from there on as positional parameters. I do not know if I can simply replace the last block in `processArguments`, because there seems to be quite some logic attached to positional parameters which I do not know.

**Comment by remkop** [comment]
No problem. I'll take a look later. Sounds like you were on the right track though. Did you push your branch so I can see it?

**Comment by RobertZenz** [comment]
I just did, you can find it in the [greedy-option branch](https://github.com/RobertZenz/picocli/tree/greedy-option). I've added a todo in `processArguments` at the location at which I believe the logic must be changed.

**Comment by remkop** [comment]
By the way, I've started to think that perhaps your desired behaviour should be the default, rather than the current behaviour.

I've also started to realize that my interpretation of `arity` has been incorrect: 
* arity should be the number of values that immediately follow the option name (not the max number of values that end up being stored in the collection field)
* the default arity for a collection/array should be 1

Redefining things like this gives the behaviour you are looking for out of the box. This means that a separate `greedy` attribute is no longer necessary, one can just use `arity = "1..*"` to get greedy parsing.

**Comment by RobertZenz** [comment]
Works either way for me, if you're ready/willing to change the default behavior, that is fine by me, as long as I get the behavior I want ;P .

 > * arity should be the number of values that immediately follow the option name (not the max number of values that end up being stored in the collection field)
 > * the default arity for a collection/array should be 1

Sounds reasonable. Out of curiosity, this will also apply to the `split` behavior you've pointed out earlier?

If you're planning on changing the default behavior, I'd advice to get a few long-term users into this discussion, though, it's always good to double check what might be going to break by such a change.

**Comment by remkop** [comment]
Some ideas to reduce disruption for existing users:

* I raised separate tickets #191 and #192 with clear titles explaining what changes are coming.
* Put a notice at the top of the http://github.com/remkop/picocli project page and the http://picocli.info manual pages
* The next version will have number 2.0 to indicate it has an incompatible change
* Write a blog post detailing these changes and post it on dzone 

I hope this will reduce disruption. Other ideas are welcome.

**Comment by remkop** [comment]
Fixes for #191, #192 and #193 are now on master. Please verify and close if this solves the problem.
(I still need to update the user manual. [Release notes](https://github.com/remkop/picocli/blob/master/RELEASE-NOTES.md) explain some of our discussions.)

Thanks again for raising this!

**Comment by RobertZenz** [comment]
I've just build the latest master (39f1944) and it works, thank you very very much.
